### DISCUSSION TITLE
builds.config metadata is undocumented

### DISCUSSION THREAD

**Discussion by jakobrosenberg** [vercel/vercel#5089] [Help]
> config (Object): Optionally, an object including arbitrary metadata to be passed to the Builder.
https://vercel.com/docs/configuration#project/builds

Haven't been able to find any documentation on metadata props like `distDir`.

**Comment by styfle** [comment]
There is no documentation on `config` because it depends on `use`. Each Runtime can implement it differently.

Can you share your `vercel.json`? You can likely remove `builds` and `vercel.json` completely using [zero config](https://vercel.com/blog/zero-config) deployments.

**Comment by jakobrosenberg** [comment]
@styfle thanks for your reply.

I'm building a framework which provides automatic deployment to services like Vercel. To avoid polluting the root folder I keep functions in a separate folder, hence the `builds` config.

The runtime I can't find documentation for is `@vercel/static-build`. With `vercel dev` I can access files in the root of `dist`, not but those in nested folders.

```
dist/
  hello.html     //accessible
  folder/
    noluck.html    //not accessible
```

vercel.json
```json
{
  "version": 2,
  "builds": [
    {
      "src": "package.json",
      "use": "@vercel/static-build",
      "config": {"distDir": "dist"}
    },
    {
      "src": "_roxi/.roxi/*.js",
      "use": "@vercel/node",
      "config": { "includeFiles": "dist/**" }
    }
  ]
}
```

**Comment by styfle** [comment]
Is this just affecting `vercel dev` or does the issue affect real deployments too?

I would suggest avoiding the `builds` property if you can and instead use zero config defaults:

- rename `/_roxi/.roxi` to `/api` for your node functions
- rename `/dist` to `/public` for your static assets

**Comment by jakobrosenberg** [comment]
This is only a problem in `vercel dev`.

Zero config isn't technically possible for us, even if we wanted, since we rely on other libraries like `vite` where folders are hardcoded.

Is there a method for providing framework functions without hardcoding them directly in the starter project?

**Comment by styfle** [comment]
You can try [`functions`](https://vercel.com/docs/configuration#project/functions) with the `_roxi/.roxi/*.js` pattern

**Comment by jakobrosenberg** [comment]
@styfle I've tried that, but IIRC you can't combine `functions` and `builds` in the same config and `functions` alone doesn't work with functions outside the `api` folder.

**Comment by styfle** [comment]
I think the problem with your example is that you have two separate builds trying to use the same directory `dist`.

The `@vercel/node` input doesn't know about the `@vercel/static-build` output because both builds happen concurrently.

If you need a build step for `@vercel/node` then you'll need to use [`vercel-build`](https://vercel.com/docs/runtimes#advanced-usage/advanced-node-js-usage/custom-build-step-for-node-js) script in package.json.

**Comment by styfle** [comment]
If you need more customized behavior at build and runtime, you can [create a Custom Runtime](https://github.com/vercel/vercel/blob/master/DEVELOPING_A_RUNTIME.md), publish to npm, and use it with your `builds` or `functions` properties.
